controller/framework test is flaky

From shippable, I noticed there are several failed cases, for example, 

https://app.shippable.com/builds/552bfb376206490c009730c3

E0413 17:34:38.400974   17931 reflector.go:158] watch of *api.Pod ended with: Stop requested
--- FAIL: Example (0.61s)
got:

want:
a-hello
b-controller
c-framework
FAIL
coverage: 85.9% of statements
FAIL	github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework	7.652s

cc/ @lavalamp

I'm seeing this from an unrelated change

```
E0413 16:50:08.658895   18560 reflector.go:158] watch of *api.Pod ended with: Stop requested
--- FAIL: TestUpdate (2.02 seconds)
	controller_test.go:317: observed transition "to" -> "from" for a-0-0
FAIL
FAIL	github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework	3.220s
```

--- a/pkg/controller/framework/controller_test.go
+++ b/pkg/controller/framework/controller_test.go
@@ -45,8 +45,7 @@ func Example() {
 	fifo := cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, nil, downstream)
 
 	// Let's do threadsafe output to get predictable test results.
-	outputSetLock := sync.Mutex{}
-	outputSet := util.StringSet{}
+	deletionCounter := make(chan string, 1000)
 
 	cfg := &framework.Config{
 		Queue:            fifo,
@@ -68,6 +67,7 @@ func Example() {
 					return err
 				}
 
+				// Delete this object.
 				source.Delete(newest.Object.(runtime.Object))
 			} else {
 				// Update our downstream store.
@@ -83,10 +83,8 @@ func Example() {
 					return err
 				}
 
-				// Record some output.
-				outputSetLock.Lock()
-				defer outputSetLock.Unlock()
-				outputSet.Insert(key)
+				// Report this deletion.
+				deletionCounter <- key
 			}
 			return nil
 		},
@@ -94,20 +92,23 @@ func Example() {
 
 	// Create the controller and run it until we close stop.
 	stop := make(chan struct{})
+	defer close(stop)
 	go framework.New(cfg).Run(stop)
 
 	// Let's add a few objects to the source.
-	for _, name := range []string{"a-hello", "b-controller", "c-framework"} {
+	testIDs := []string{"a-hello", "b-controller", "c-framework"}
+	for _, name := range testIDs {
 		// Note that these pods are not valid-- the fake source doesn't
 		// call validation or anything.
 		source.Add(&api.Pod{ObjectMeta: api.ObjectMeta{Name: name}})
 	}
 
 	// Let's wait for the controller to process the things we just added.
-	time.Sleep(500 * time.Millisecond)
-	close(stop)
+	outputSet := util.StringSet{}
+	for i := 0; i < len(testIDs); i++ {
+		outputSet.Insert(<-deletionCounter)
+	}
 
-	outputSetLock.Lock()
 	for _, key := range outputSet.List() {
 		fmt.Println(key)
 	}
@@ -122,8 +123,7 @@ func ExampleInformer() {
 	source := framework.NewFakeControllerSource()
 
 	// Let's do threadsafe output to get predictable test results.
-	outputSetLock := sync.Mutex{}
-	outputSet := util.StringSet{}
+	deletionCounter := make(chan string, 1000)
 
 	// Make a controller that immediately deletes anything added to it, and
 	// logs anything deleted.
@@ -141,30 +141,31 @@ func ExampleInformer() {
 					key = "oops something went wrong with the key"
 				}
 
-				// Record some output when items are deleted.
-				outputSetLock.Lock()
-				defer outputSetLock.Unlock()
-				outputSet.Insert(key)
+				// Report this deletion.
+				deletionCounter <- key
 			},
 		},
 	)
 
 	// Run the controller and run it until we close stop.
 	stop := make(chan struct{})
+	defer close(stop)
 	go controller.Run(stop)
 
 	// Let's add a few objects to the source.
-	for _, name := range []string{"a-hello", "b-controller", "c-framework"} {
+	testIDs := []string{"a-hello", "b-controller", "c-framework"}
+	for _, name := range testIDs {
 		// Note that these pods are not valid-- the fake source doesn't
-		// call validation or perform any other checking.
+		// call validation or anything.
 		source.Add(&api.Pod{ObjectMeta: api.ObjectMeta{Name: name}})
 	}
 
 	// Let's wait for the controller to process the things we just added.
-	time.Sleep(500 * time.Millisecond)
-	close(stop)
+	outputSet := util.StringSet{}
+	for i := 0; i < len(testIDs); i++ {
+		outputSet.Insert(<-deletionCounter)
+	}
 
-	outputSetLock.Lock()
 	for _, key := range outputSet.List() {
 		fmt.Println(key)
 	}

